src/modules/Game.class.js

```diff
@@ -1,68 +1,250 @@
+    this.status = Game.STATUS.idle;
+  }
+
+  resetState() {
+    this.state = this.initialState.map(row => [...row]);
+    this.score = 0;
+  }
+
+  addCells(count = 1) {
+    for (let i = 0; i < count; i++) {
+      this.addNewTile();
+    }
+
+    if (this.isVictory()) {
+      this.status = Game.STATUS.win;
+    } else if (!this.validState()) {
+      this.status = Game.STATUS.lose;
+    }
+  }
+
+  addNewTile() {
+    const emptyCells = this.getEmptyCells();
+
+    if (!emptyCells.length) {
+      return;
+    }
+
+    const randomCell = Math.floor(Math.random() * emptyCells.length);
+    const [row, col] = emptyCells[randomCell];
+
+    this.state[row][col] = Math.random() < 0.9 ? 2 : 4;
+  }
+
+  getEmptyCells() {
+    return this.state
+      .flatMap((row, rowIndex) =>
+        row.map((cell, colIndex) => (cell === 0
+          ? [rowIndex, colIndex]
+          : null
+        )),
+      )
+      .filter(cell => cell !== null);
+  }
+
+  isVictory() {
+    return this.state.flat().some(tile => tile === 2048);
+  }
+
+  validState() {
+    const size = this.state.length;
+
+    for (let row = 0; row < size; row++) {
+      for (let col = 0; col < size; col++) {
+        const current = this.state[row][col];
+
+        if (
+          current === 0
+          || (col < size - 1 && current === this.state[row][col + 1])
+          || (row < size - 1 && current === this.state[row + 1][col])
+        ) {
+          return true;
+        }
+      }
+    }
+
+    return false;
+  }
+
+  moveLeft() {
+    if (this.status !== Game.STATUS.playing) {
+      return;
+    }
+
+    const updatedState = this.state.map(row => this.applyMove(row));
+
+    if (!this.isStateEqual(this.state, updatedState)) {
+      this.state = updatedState;
+      this.addCells();
+    }
+  }
+
+  moveRight() {
+    if (this.status !== Game.STATUS.playing) {
+      return;
+    }
+
+    const reversedState = this.state.map(row => [...row].reverse());
+    const updatedState = reversedState.map(row =>
+      this.applyMove(row).reverse(),
+    );
+
+    if (!this.isStateEqual(this.state, updatedState)) {
+      this.state = updatedState;
+      this.addCells();
+    }
+  }
+
+  moveUp() {
+    if (this.status !== Game.STATUS.playing) {
+      return;
+    }
+
+    const rotatedState = this.rotateRight(this.state);
+    const updatedState = rotatedState.map(row => this.applyMove(row));
+    const unrotatedState = this.rotateLeft(updatedState);
+
+    if (!this.isStateEqual(this.state, unrotatedState)) {
+      this.state = unrotatedState;
+      this.addCells();
+    }
+  }
+
+  moveDown() {
+    if (this.status !== Game.STATUS.playing) {
+      return;
+    }
+
+    const rotatedState = this.rotateLeft(this.state);
+    const updatedState = rotatedState.map(row => this.applyMove(row));
+    const unrotatedState = this.rotateRight(updatedState);
+
+    if (!this.isStateEqual(this.state, unrotatedState)) {
+      this.state = unrotatedState;
+      this.addCells();
```

Comment:
you can move these lines to handleKeydown function and write it only one time there